crates/core/executor/src/hook.rs

```diff
@@ -204,6 +208,202 @@ mod ecrecover {
     }
 }
 
+mod fp_ops {
+    use super::{pad_to_be, BigUint, HookEnv, One, Zero};
+
+    pub fn hook_fp_inverse(_: HookEnv, buf: &[u8]) -> Vec<Vec<u8>> {
+        let len: usize = u32::from_be_bytes(buf[0..4].try_into().unwrap()) as usize;
+
+        assert!(buf.len() == 4 + 2 * len, "FpOp: Invalid buffer length");
+
+        let buf = &buf[4..];
+        let element = BigUint::from_bytes_be(&buf[..len]);
+        let modulus = BigUint::from_bytes_be(&buf[len..2 * len]);
+
+        assert!(!element.is_zero(), "FpOp: Inverse called with zero");
+
+        let inverse = element.modpow(&(&modulus - BigUint::from(2u64)), &modulus);
+
+        vec![pad_to_be(&inverse, len)]
+    }
+
+    /// Compute the square root of a field element.
+    ///
+    /// # Arguments
+    /// * `env` - The environment in which the hook is invoked.
+    /// * `buf` - The buffer containing the data needed to compute the square root.
+    ///     - [ len || Element || Modulus || NQR ]
+    ///     - len is the length of the element, modulus, and nqr in Big endian.
+    ///     - Element is the field element to compute the square root of, interpreted as a big endian integer of `len` bytes.
+    ///     - Modulus is the modulus of the field, interpreted as a big endian integer of `len` bytes.
+    ///     - NQR is the non-quadratic residue of the field, interpreted as a big endian integer of `len` bytes.
+    ///
+    /// # Assumptions
+    /// - NQR is a non-quadratic residue of the field.
+    ///
+    /// The result is a single 32 byte vector containing the square root.
+    pub fn hook_fp_sqrt(_: HookEnv, buf: &[u8]) -> Vec<Vec<u8>> {
+        let len: usize = u32::from_be_bytes(buf[0..4].try_into().unwrap()) as usize;
+
+        assert!(buf.len() == 4 + 3 * len, "FpOp: Invalid buffer length");
+
+        let buf = &buf[4..];
+        let element = BigUint::from_bytes_be(&buf[..len]);
+        let modulus = BigUint::from_bytes_be(&buf[len..2 * len]);
+        let nqr = BigUint::from_bytes_be(&buf[2 * len..3 * len]);
+
+        assert!(
+            element < modulus,
+            "Element is not less than modulus, the hook only accepts canonical representations"
+        );
+        assert!(
+            nqr < modulus,
+            "NQR is zero or non-canonical, the hook only accepts canonical representations"
+        );
+
+        // The sqrt of zero is zero.
+        if element.is_zero() {
+            return vec![vec![1], vec![0; len]];
+        }
+
+        // We compute the square root of the element using the general tonelli shanks algorithm.
```

Comment:
No first-person tense.

"Compute the square root of the element using the general Tonelli-Shanks algorithm."